File: app/src/main/java/eu/rtsketo/sakurastats/hashmaps/SDPMap.java
package eu.rtsketo.sakurastats.hashmaps;

import android.content.res.Resources;
import android.util.SparseIntArray;

import eu.rtsketo.sakurastats.BuildConfig;

public class SDPMap {
    private static SparseIntArray sdpMap;
    private static Resources res;

    public static int sdp2px(int sdp) {
        if (sdp < 1) return -1;
        if (sdpMap == null) sdpMap = new SparseIntArray();
        int px = sdpMap.get(sdp, 1337);
        if (px == 1337) {
            String pack = BuildConfig.APPLICATION_ID;
            int id = res.getIdentifier("_" +
                    sdp + "sdp", "dimen", pack);
            px = res.getDimensionPixelSize(id);
            sdpMap.put(sdp, px); }
        return px; }

    public static int dp2px(float dp) {
        return Math.round(dp * Resources.getSystem().getDisplayMetrics().density); }

    public static float px2dp(float px) {
        return px / Resources.getSystem().getDisplayMetrics().density; }

    public static float sdp2dp(int sdp) {
        return px2dp(sdp2px(sdp)); }

    public int getScreenWidth() {
        return Resources.getSystem().getDisplayMetrics().widthPixels; }

    public int getScreenHeight() {
        return Resources.getSystem().getDisplayMetrics().heightPixels; }

    public static void init(Resources res) { SDPMap.res = res; }
}

File: app/src/main/java/eu/rtsketo/sakurastats/hashmaps/SDPMap.kt
package eu.rtsketo.sakurastats.hashmaps

import android.content.res.Resources
import android.util.SparseIntArray
import androidx.multidex.BuildConfig
import kotlin.math.roundToInt

class SDPMap {
    val screenWidth: Int
        get() = Resources.getSystem().displayMetrics.widthPixels

    val screenHeight: Int
        get() = Resources.getSystem().displayMetrics.heightPixels

    companion object {
        private var sdpMap = SparseIntArray()
        private lateinit var res: Resources

        fun sdp2px(sdp: Int): Int {
            if (sdp < 1) return -1
            var px = sdpMap[sdp, 1337]

            if (px == 1337) {
                val pack = BuildConfig.APPLICATION_ID
                val id = res.getIdentifier("_" +
                        sdp + "sdp", "dimen", pack)
                px = res.getDimensionPixelSize(id)
                sdpMap.put(sdp, px)
            }
            return px
        }

        fun dp2px(dp: Float): Int {
            return (dp * Resources.getSystem().displayMetrics.density).roundToInt()
        }

        fun px2dp(px: Float): Float {
            return px / Resources.getSystem().displayMetrics.density
        }

        fun sdp2dp(sdp: Int): Float {
            return px2dp(sdp2px(sdp).toFloat())
        }

        fun init(res: Resources) {
            this.res = res
        }
    }
}